AMICI installation on windows

I've been trying to install amici on a windows machine using Anaconda. I followed the instructions as described [here](https://amici.readthedocs.io/en/latest/python_installation.html) to install MinGW, OpenBLAS and SWIG, and added the suggested environment variables. Then I created a new conda environment and used the following command to install amici:

pip install --global-option="build_clib" --global-option="--compiler=mingw32" --global-option="build_ext" --global-option="--compiler=ming32" amici --no-cache-dir --verbose

But installation fails with a rather elaborate error message. Is there anything I should have done differently?

[amici_error_log1.txt](https://github.com/AMICI-dev/AMICI/files/6446603/amici_error_log1.txt)

I can see the following output at the end of the error log:
`  C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iamici/include -Iamici/ThirdParty/SuiteSparse/KLU/Include/ -Iamici/ThirdParty/SuiteSparse/AMD/Include/ -Iamici/ThirdParty/SuiteSparse/COLAMD/Include/ -Iamici/ThirdParty/SuiteSparse/BTF/Include/ -Iamici/ThirdParty/SuiteSparse/SuiteSparse_config -Iamici/ThirdParty/SuiteSparse/include -Iamici/ThirdParty/sundials/include -Iamici/ThirdParty/sundials/src -Iamici/ThirdParty/gsl/ -IC:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\include -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt -IC:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt /EHsc /Tpamici\src\cblas.cpp /Fobuild\temp.win-amd64-3.9\amici\src\cblas.obj /IC:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10 /std:c++14
  cblas.cpp
  amici\src\cblas.cpp(16): fatal error C1083: Cannot open include file: 'cblas.h': No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29910\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2`

The entry
`/IC:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10`
shows me that you have created the environment variable that tells the compiler where to find cblas.h, but it is not finding it. It looks like something is missing in your installation of openBLAS. Does the directory `C:\BLAS\OpenBLAS-v0.3.10\OpenBLAS-0.3.10` contain a copy of cblas.h?

The entry
` C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe`
shows me that the python distutils are using MSVC (Microsoft Visual Studio) to compile and link the amici module cblas.cpp (which requires cblas.h), but I thought you were using mingw32.

Alright, I've removed all traces of Anaconda from my system, built BLAS and added the environment variables (BLAS_CFLAGS and BLAS_LIBS), installed SWIG and added to path but somehow still end up with the same error when trying to install AMICI:

  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemv
  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemm
  amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_daxpy
  build\lib.win-amd64-3.8\amici\_amici.cp38-win_amd64.pyd : fatal error LNK1120: 3 unresolved externals
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.28.29910\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120

Is there any other environment variables that I might be missing?

[error_log_full.txt](https://github.com/AMICI-dev/AMICI/files/6467987/error_log_full.txt)

> Alright, I've removed all traces of Anaconda from my system, built BLAS and added the environment variables (BLAS_CFLAGS and BLAS_LIBS), installed SWIG and added to path but somehow still end up with the same error when trying to install AMICI:
> 
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemv
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_dgemm
> amici.lib(cblas.obj) : error LNK2001: unresolved external symbol cblas_daxpy
> build\lib.win-amd64-3.8\amici_amici.cp38-win_amd64.pyd : fatal error LNK1120: 3 unresolved externals
> error: command 'C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\link.exe' failed with exit status 1120
> 
> Is there any other environment variables that I might be missing?
> 
> [error_log_full.txt](https://github.com/AMICI-dev/AMICI/files/6467987/error_log_full.txt)

No there aren't any additional environment variables that need to be set. Comparing to https://github.com/AMICI-dev/AMICI/runs/2450686166?check_suite_focus=true , the linker invocation looks correct. Two possible sources of error that come to mind:

- openBLAS and AMICI need to be compiled with exactly the same compiler, are you sure this is the case?
- can you confirm there is a file `openblas.lib` located in `C:\BLAS\lib`?

If both are true then we are probably left to inspecting the symbol tables themselves to figure out where things fails. I have never done this on windows, but https://stackoverflow.com/questions/305287/how-to-see-the-contents-of-windows-library-lib looks like a good resource. The libraries you want to look at are are `C:\BLAS\lib\openblas.lib` and `amici.lib`, which should be located in `./build/` in the location where you are installing amici. Its easier to track this directory down when using an in-source installation as explained in https://amici.readthedocs.io/en/latest/python_installation.html#installation-of-development-versions.

--- a/documentation/python_installation.rst
+++ b/documentation/python_installation.rst
@@ -266,6 +266,18 @@ by MSVC (see Visual Studio above). ``KERNEL32.dll`` is part of Windows and in
 ``C:\Windows\System32``. The ``api-ms-win-crt-XXX-l1-1-0.dll`` are needed by
 ``openblas.dll`` and are part of the Windows Universal C Runtime.
 
+.. note::
+
+    Since Python 3.8, the library directory needs to be set as follows:
+
+    .. code-block:: python
+
+        import os
+        # directory containing `openblas.dll`
+        os.add_dll_directory("C:\\BLAS\\bin")
+        import amici
+
+    Adding it to ``PATH`` will not work.
 
 Further topics
 ++++++++++++++

